WalletWasabi/Services/UpdateManager.cs

```diff
@@ -0,0 +1,238 @@
+using Newtonsoft.Json.Linq;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+using WalletWasabi.Microservices;
+using WalletWasabi.Models;
+using WalletWasabi.Tor.Http;
+
+namespace WalletWasabi.Services;
+
+public class UpdateManager
+{
+	private string InstallerName { get; set; } = "";
+	private const byte MaxTries = 3;
+	private const string ReleaseURL = "https://api.github.com/repos/zkSNACKs/WalletWasabi/releases/latest";
+
+	public UpdateManager(string dataDir, bool downloadNewVersion, IHttpClient httpClient)
+	{
+		DownloadsDir = Path.Combine(dataDir, "Downloads");
+		HttpClient = httpClient;
+		DownloadNewVersion = downloadNewVersion;
+	}
+
+	public async void UpdateStatusChangedAsync(UpdateStatus updateStatus)
+	{
+		var tries = 0;
+		bool updateAvailable = !updateStatus.ClientUpToDate || !updateStatus.BackendCompatible;
+		Version targetVersion = updateStatus.ClientVersion;
+		if (!updateAvailable)
+		{
+			return;
+		}
+		if (DownloadNewVersion)
+		{
+			Logger.LogInfo($"Trying to download new version: {targetVersion}");
+			do
+			{
+				tries++;
+				try
+				{
+					bool isReadyToInstall = await GetInstallerAsync(targetVersion).ConfigureAwait(false);
+					Logger.LogInfo($"Version {targetVersion} downloaded successfuly.");
+					updateStatus.IsReadyToInstall = isReadyToInstall;
+					break;
+				}
+				catch (Exception ex)
+				{
+					Logger.LogError($"Geting version {targetVersion} failed with error.", ex);
+				}
+			} while (tries < MaxTries);
+		}
+
+		UpdateAvailableToGet?.Invoke(this, updateStatus);
+	}
+
+	private async Task<bool> GetInstallerAsync(Version targetVersion)
+	{
+		if (CheckIfInstallerDownloaded())
+		{
+			return true;
+		}
+		using HttpRequestMessage message = new(HttpMethod.Get, ReleaseURL);
+		message.Headers.UserAgent.Add(new("WalletWasabi", "2.0"));
+		var response = await HttpClient.SendAsync(message).ConfigureAwait(false);
+
+		JObject jsonResponse = JObject.Parse(await response.Content.ReadAsStringAsync().ConfigureAwait(false));
+		string softwareVersion = jsonResponse["tag_name"]?.ToString() ?? throw new InvalidDataException("Endpoint gave back wrong json data or it's changed.");
```

Comment:
Fix nullables

Reply:
There are no nullable warnings.

Reply:
Maybe I reviewed an outdated version. It is OK!